worker/download.go

```diff
@@ -0,0 +1,1088 @@
+		mgr *downloadManager
+
+		dID       id
+		sID       slabID
+		created   time.Time
+		index     int
+		minShards int
+		length    uint32
+		offset    uint32
+
+		mu            sync.Mutex
+		lastOverdrive time.Time
+		numInflight   uint64
+		numLaunched   uint64
+		numCompleted  int
+
+		curr          types.PublicKey
+		hostToSectors map[types.PublicKey][]sectorInfo
+		used          map[types.PublicKey]struct{}
+
+		sectors [][]byte
+		errs    HostErrorSet
+	}
+
+	slabDownloadResponse struct {
+		shards [][]byte
+		index  int
+		err    error
+	}
+
+	sectorDownloadReq struct {
+		ctx context.Context
+
+		length uint32
+		offset uint32
+		root   types.Hash256
+		hk     types.PublicKey
+
+		overdrive    bool
+		sectorIndex  int
+		responseChan chan sectorDownloadResp
+	}
+
+	sectorDownloadResp struct {
+		hk          types.PublicKey
+		sectorIndex int
+		sector      []byte
+		err         error
+	}
+
+	sectorInfo struct {
+		object.Sector
+		index int
+	}
+
+	downloadManagerStats struct {
+		avgDownloadSpeedMBPS float64
+		avgOverdrivePct      float64
+		downloaders          map[types.PublicKey]downloaderStats
+	}
+)
+
+func (w *worker) initDownloadManager(maxOverdrive uint64, overdriveTimeout time.Duration, logger *zap.SugaredLogger) {
+	if w.downloadManager != nil {
+		panic("download manager already initialized") // developer error
+	}
+
+	w.downloadManager = newDownloadManager(w, maxOverdrive, overdriveTimeout, logger)
+}
+
+func newDownloadManager(hp hostProvider, maxOverdrive uint64, overdriveTimeout time.Duration, logger *zap.SugaredLogger) *downloadManager {
+	return &downloadManager{
+		hp:     hp,
+		logger: logger,
+
+		maxOverdrive:     maxOverdrive,
+		overdriveTimeout: overdriveTimeout,
+
+		statsOverdrivePct:                newDataPoints(0),
+		statsSlabDownloadSpeedBytesPerMS: newDataPoints(0),
+
+		stopChan: make(chan struct{}),
+
+		ongoing:     make(map[slabID]struct{}),
+		downloaders: make(map[types.PublicKey]*downloader),
+	}
+}
+
+func (mgr *downloadManager) newDownloader(host hostV3) *downloader {
+	return &downloader{
+		host: host,
+
+		statsSectorDownloadEstimateInMS: newDataPoints(statsDecayHalfTime),
+		statsDownloadSpeedBytesPerMS:    newDataPoints(0), // no decay for exposed stats
+
+		signalWorkChan: make(chan struct{}, 1),
+		stopChan:       make(chan struct{}),
+
+		queue: make([]*sectorDownloadReq, 0),
+	}
+}
+
+func (mgr *downloadManager) DownloadObject(ctx context.Context, w io.Writer, o object.Object, offset, length uint32, contracts []api.ContractMetadata) (err error) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "download")
+	defer func() {
+		span.RecordError(err)
+		span.End()
+	}()
+
+	// calculate what slabs we need
+	slabs := slabsForDownload(o.Slabs, offset, length)
+	if len(slabs) == 0 {
+		return nil
+	}
+
+	// ensure everything cancels if download is done
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// refresh the downloaders
+	mgr.refreshDownloaders(contracts)
```

Comment:
Does it make sense to refresh the downloaders using contracts instead of passing funded accounts? Since we don't use the contracts here but accounts?

Reply:
Yeah I considered it when writing it, the problem is that you either have to refactor `hostV3` or pass in accounts but you still need a price table. I figured it wasn't worth tackling in a first iteration, but if we decide we want to do it just follow up with a dedicated refactor.

It's simply because you need a `fcid` to construct a `hostV3` but the download manager won't ever use it. Splitting up `hostV3` is a little gnarly though, I considered having an `eahost` and `contractHost` or something but I'm not sure if it's worth it.